lib/formContext.ts

```diff
@@ -325,6 +325,255 @@ export const removeChoiceFromRules = (elements: FormElement[], choiceId: string)
+    return false;
+  });
+};
+
+/**
+ * Get all rules that are within the parent choiceId
+ * @param rules - The rules to filter
+ * @param choiceId - The choiceId to filter by
+ * @returns {Array} - Returns an array of rules that are within the parent choiceId
+ */
+export const filterRulesWithChoiceIdParent = (rules: ConditionalRule[], choiceId: string) => {
+  return rules.filter((rule) => {
+    const choiceParentId = choiceId.split(".")[0];
+    const ruleChoiceParentId = rule.choiceId.split(".")[0];
+
+    if (choiceParentId === ruleChoiceParentId) {
+      return true;
+    }
+    return false;
+  });
+};
+
+/**
+ * Decrement the index of the choiceId
+ * @param rules - The rules to adjust
+ * @returns {Array} - Returns an array of rules with the choiceId index decremented
+ */
+const decrementChoiceIdIndexes = (rules: ConditionalRule[]) => {
+  // Subtract 1 from the index of the choiceId
+  const adjustedRules = rules
+    .map((rule) => {
+      const choiceParts = rule.choiceId.split(".");
+      const choiceIndex = Number(choiceParts[1]);
+      const newIndex = Number(choiceIndex) - 1;
+      const newChoiceId = `${choiceParts[0]}.${newIndex}`;
+
+      if (newIndex >= 0) {
+        return { choiceId: newChoiceId };
+      }
+
+      return { choiceId: "" };
+    })
+    .filter((rule) => rule?.choiceId !== "");
+
+  return adjustedRules;
+};
+
+/**
+ * Filter out rules that are below the choiceId index
+ * @param rules - The rules to filter
+ * @param choiceId - The choiceId to filter by
+ * @returns {Array} - Returns an array of rules that are above the choiceId index
+ * i.e. if the choiceId is "1.3" and the rule is "1.0", "1.1" or  "1.2" etc...
+ * we want to leave "1.0", "1.1" and "1.2" out of the array we're going to adjust
+ */
+export const filterOutRulesBelowChoiceId = (rules: ConditionalRule[], choiceId: string) => {
+  return rules.filter((rule) => {
+    const choiceIndex = Number(choiceId.split(".")[1]);
+    const ruleIndex = Number(rule.choiceId.split(".")[1]);
+    if (ruleIndex > choiceIndex) {
+      return true;
+    }
+    return false;
+  });
+};
+
+/**
+ * Convert an array of rules to an array of choiceIds
+ * @param rules - The rules to convert
+ * @returns {Array} - Returns an array of choiceIds
+ * i.e. [{ choiceId: "1.0" }, { choiceId: "1.1" }] becomes ["1.0", "1.1"]
+ */
+const toChoiceValues = (rules: ConditionalRule[]) => {
+  return rules.map((rule) => rule.choiceId);
+};
+
+/**
+ * Find the difference between two arrays
+ * @param insideRange - The array to compare
+ * @param aboveRange - The array to compare
+ * @returns {Array} - Returns an array of the difference between the two arrays
+ */
+const diffAboveRange = (insideRange: ConditionalRule[], aboveRange: ConditionalRule[]) => {
+  const insideRangeSet = new Set(toChoiceValues(insideRange));
+  const aboveRangeSet = new Set(toChoiceValues(aboveRange));
+  const rangeDifference = new Set([...insideRangeSet].filter((x) => !aboveRangeSet.has(x)));
```

Comment:
Nice

Reply:
Was hoping to use https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/difference

But can't yet